# 🧹 Car Insurance Data Cleaning

This notebook is focused solely on cleaning the `MachineLearningRating_v3.txt` dataset. Cleaning steps include handling missing values, fixing incorrect data types, and removing duplicates.


# Step 1: Import Libraries

In this step, we import the necessary Python libraries.

In [3]:
import pandas as pd
import numpy as np


## 📂 Load Dataset

The dataset is stored in the `data/` directory. It is a pipe-delimited `.txt` file.


In [7]:
file_path = '../data/MachineLearningRating_v3.txt'
df = pd.read_csv(file_path, delimiter='|', low_memory=False)

# Preview
print(f"Shape: {df.shape}")
df.head()


Shape: (1000098, 52)


UnderwrittenCoverID  PolicyID     TransactionMonth  IsVATRegistered  \
0               145249     12827  2015-03-01 00:00:00             True   
1               145249     12827  2015-05-01 00:00:00             True   
2               145249     12827  2015-07-01 00:00:00             True   
3               145255     12827  2015-05-01 00:00:00             True   
4               145255     12827  2015-07-01 00:00:00             True   

  Citizenship          LegalType Title Language                 Bank  \
0              Close Corporation    Mr  English  First National Bank   
1              Close Corporation    Mr  English  First National Bank   
2              Close Corporation    Mr  English  First National Bank   
3              Close Corporation    Mr  English  First National Bank   
4              Close Corporation    Mr  English  First National Bank   

       AccountType  ...                    ExcessSelected CoverCategory  \
0  Current account  ...             Mobility - Windscreen    Windscreen   
1  Current account  ...             Mobility - Windscreen    Windscreen   
2  Current account  ...             Mobility - Windscreen    Windscreen   
3  Current account  ...  Mobility - Metered Taxis - R2000    Own damage   
4  Current account  ...  Mobility - Metered Taxis - R2000    Own damage   

    CoverType            CoverGroup              Section  \
0  Windscreen  Comprehensive - Taxi  Motor Comprehensive   
1  Windscreen  Comprehensive - Taxi  Motor Comprehensive   
2  Windscreen  Comprehensive - Taxi  Motor Comprehensive   
3  Own Damage  Comprehensive - Taxi  Motor Comprehensive   
4  Own Damage  Comprehensive - Taxi  Motor Comprehensive   

                           Product StatutoryClass StatutoryRiskType  \
0  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   
1  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   
2  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   
3  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   
4  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   

   TotalPremium TotalClaims  
0     21.929825         0.0  
1     21.929825         0.0  
2      0.000000         0.0  
3    512.848070         0.0  
4      0.000000         0.0  

[5 rows x 52 columns]

## 🧾 Data Overview

We'll inspect the data types and structure.


In [8]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000098 entries, 0 to 1000097
Data columns (total 52 columns):
 #   Column                    Non-Null Count    Dtype  
---  ------                    --------------    -----  
 0   UnderwrittenCoverID       1000098 non-null  int64  
 1   PolicyID                  1000098 non-null  int64  
 2   TransactionMonth          1000098 non-null  object 
 3   IsVATRegistered           1000098 non-null  bool   
 4   Citizenship               1000098 non-null  object 
 5   LegalType                 1000098 non-null  object 
 6   Title                     1000098 non-null  object 
 7   Language                  1000098 non-null  object 
 8   Bank                      854137 non-null   object 
 9   AccountType               959866 non-null   object 
 10  MaritalStatus             991839 non-null   object 
 11  Gender                    990562 non-null   object 
 12  Country                   1000098 non-null  object 
 13  Province                  1

## 🚫 Missing Values

We'll identify and handle missing values in the dataset.


In [9]:
# Check missing values
missing = df.isnull().sum()
missing[missing > 0]


Bank                        145961
AccountType                  40232
MaritalStatus                 8259
Gender                        9536
mmcode                         552
VehicleType                    552
make                           552
Model                          552
Cylinders                      552
cubiccapacity                  552
kilowatts                      552
bodytype                       552
NumberOfDoors                  552
VehicleIntroDate               552
CustomValueEstimate         779642
CapitalOutstanding               2
NewVehicle                  153295
WrittenOff                  641901
Rebuilt                     641901
Converted                   641901
CrossBorder                 999400
NumberOfVehiclesInFleet    1000098
dtype: int64

## 🛠 Fix Data Types

Convert object columns to proper data types (e.g., dates, categories, numbers).


In [11]:
df['Gender'] = df['Gender'].astype('category')


## ♻️ Remove Duplicates

Check and remove duplicate rows.


In [12]:
print("Duplicate rows:", df.duplicated().sum())
df = df.drop_duplicates()
print("New shape:", df.shape)


Duplicate rows: 0
New shape: (1000098, 52)


In [13]:
# Automatically categorize columns based on their data types
categorical_cols = df.select_dtypes(include=['object', 'category']).columns.tolist()
numerical_cols = df.select_dtypes(include=['number']).columns.tolist()

print("Categorical columns:")
print(categorical_cols)

print("\nNumerical columns:")
print(numerical_cols)


Categorical columns:
['TransactionMonth', 'Citizenship', 'LegalType', 'Title', 'Language', 'Bank', 'AccountType', 'MaritalStatus', 'Gender', 'Country', 'Province', 'MainCrestaZone', 'SubCrestaZone', 'ItemType', 'VehicleType', 'make', 'Model', 'bodytype', 'VehicleIntroDate', 'AlarmImmobiliser', 'TrackingDevice', 'CapitalOutstanding', 'NewVehicle', 'WrittenOff', 'Rebuilt', 'Converted', 'CrossBorder', 'TermFrequency', 'ExcessSelected', 'CoverCategory', 'CoverType', 'CoverGroup', 'Section', 'Product', 'StatutoryClass', 'StatutoryRiskType']

Numerical columns:
['UnderwrittenCoverID', 'PolicyID', 'PostalCode', 'mmcode', 'RegistrationYear', 'Cylinders', 'cubiccapacity', 'kilowatts', 'NumberOfDoors', 'CustomValueEstimate', 'NumberOfVehiclesInFleet', 'SumInsured', 'CalculatedPremiumPerTerm', 'TotalPremium', 'TotalClaims']


In [14]:
# Check missing values in categorical columns
missing_categorical = df[categorical_cols].isnull().sum()
missing_categorical = missing_categorical[missing_categorical > 0]

# Check missing values in numerical columns
missing_numerical = df[numerical_cols].isnull().sum()
missing_numerical = missing_numerical[missing_numerical > 0]

print("Categorical columns with missing values:")
print(missing_categorical)

print("\nNumerical columns with missing values:")
print(missing_numerical)


Categorical columns with missing values:
Bank                  145961
AccountType            40232
MaritalStatus           8259
Gender                  9536
VehicleType              552
make                     552
Model                    552
bodytype                 552
VehicleIntroDate         552
CapitalOutstanding         2
NewVehicle            153295
WrittenOff            641901
Rebuilt               641901
Converted             641901
CrossBorder           999400
dtype: int64

Numerical columns with missing values:
mmcode                         552
Cylinders                      552
cubiccapacity                  552
kilowatts                      552
NumberOfDoors                  552
CustomValueEstimate         779642
NumberOfVehiclesInFleet    1000098
dtype: int64


In [18]:
# Check only numeric columns
numeric_cols = df.select_dtypes(include=[np.number]).columns

# Find columns that are all zeros (after filling NaNs with 0)
all_zero_cols = [col for col in numeric_cols if (df[col].fillna(0) == 0).all()]

# Columns that are all NaN (for all dtypes)
all_nan_cols = [col for col in df.columns if df[col].isna().all()]

print("Columns with all zeros:", all_zero_cols)
print("Columns with all NaNs:", all_nan_cols)

# Drop both
cols_to_drop = list(set(all_zero_cols + all_nan_cols))
df.drop(columns=cols_to_drop, inplace=True)

print(" Dropped columns:", cols_to_drop)
print(" New shape:", df.shape)


Columns with all zeros: []
Columns with all NaNs: []
 Dropped columns: []
 New shape: (1000098, 51)


## 🧮 Step 6: Count Gender and Marital Status Values

Before proceeding to final checks, let’s understand how these two important categorical columns are distributed.


In [45]:
# Check if the columns exist
if 'Gender' in df.columns:
    print(" Gender counts:")
    print(df['Gender'].value_counts(dropna=False))
else:
    print(" 'Gender' column not found.")

print("\n" + "="*40 + "\n")

if 'MaritalStatus' in df.columns:
    print(" Marital Status counts:")
    print(df['MaritalStatus'].value_counts(dropna=False))
else:
    print(" 'MaritalStatus' column not found.")


 Gender counts:
Gender
Unknown    949679
Male        42468
Female       6755
Name: count, dtype: int64


 Marital Status counts:
MaritalStatus
Unknown    993271
Single       4254
Married      1377
Name: count, dtype: int64


In [43]:
# Handle Gender
if 'Gender' in df.columns and df['Gender'].dtype.name == 'category':
    if 'Unknown' not in df['Gender'].cat.categories:
        df['Gender'] = df['Gender'].cat.add_categories(['Unknown'])
    df['Gender'] = df['Gender'].fillna('Unknown')
    df['Gender'] = df['Gender'].replace('Not specified', 'Unknown')

# Handle MaritalStatus
if 'MaritalStatus' in df.columns and df['MaritalStatus'].dtype.name == 'category':
    if 'Unknown' not in df['MaritalStatus'].cat.categories:
        df['MaritalStatus'] = df['MaritalStatus'].cat.add_categories(['Unknown'])
    df['MaritalStatus'] = df['MaritalStatus'].fillna('Unknown')
    df['MaritalStatus'] = df['MaritalStatus'].replace('Not specified', 'Unknown')


In [46]:
# List all columns in the DataFrame
print(" All Columns in the DataFrame:\n")
print(df.columns.tolist())

# Show column data types
print("\n Column Data Types:\n")
print(df.dtypes)

# Show number of missing values per column (if any)
print("\n Missing Values Per Column:\n")
missing_summary = df.isna().sum()
print(missing_summary[missing_summary > 0] if missing_summary.sum() > 0 else "✅ No missing values.")


 All Columns in the DataFrame:

['UnderwrittenCoverID', 'PolicyID', 'TransactionMonth', 'IsVATRegistered', 'Citizenship', 'LegalType', 'Title', 'Bank', 'AccountType', 'MaritalStatus', 'Gender', 'Province', 'PostalCode', 'MainCrestaZone', 'SubCrestaZone', 'mmcode', 'VehicleType', 'RegistrationYear', 'make', 'Model', 'Cylinders', 'cubiccapacity', 'kilowatts', 'bodytype', 'NumberOfDoors', 'VehicleIntroDate', 'AlarmImmobiliser', 'TrackingDevice', 'CapitalOutstanding', 'NewVehicle', 'SumInsured', 'CalculatedPremiumPerTerm', 'CoverCategory', 'CoverType', 'CoverGroup', 'Section', 'Product', 'TotalPremium', 'TotalClaims']

 Column Data Types:

UnderwrittenCoverID            int64
PolicyID                       int64
TransactionMonth              object
IsVATRegistered                 bool
Citizenship                   object
LegalType                     object
Title                         object
Bank                          object
AccountType                   object
MaritalStatus          

In [25]:
# Dataset shape
print(f" Dataset contains {df.shape[0]:,} rows and {df.shape[1]} columns.\n")

# Percent of missing values
missing_info = df.isna().sum()
missing_info = missing_info[missing_info > 0].sort_values(ascending=False)
missing_percent = (missing_info / len(df)) * 100

# Display summary
final_missing_df = pd.DataFrame({
    'MissingValues': missing_info,
    'Percent': missing_percent.round(2)
})
print("Columns with Missing Data:\n")
print(final_missing_df)


 Dataset contains 1,000,098 rows and 51 columns.

Columns with Missing Data:

                     MissingValues  Percent
CrossBorder                 999400    99.93
CustomValueEstimate         779642    77.96
Rebuilt                     641901    64.18
Converted                   641901    64.18
WrittenOff                  641901    64.18
NewVehicle                  153295    15.33
Bank                        145961    14.59
AccountType                  40232     4.02
mmcode                         552     0.06
VehicleType                    552     0.06
make                           552     0.06
NumberOfDoors                  552     0.06
bodytype                       552     0.06
kilowatts                      552     0.06
cubiccapacity                  552     0.06
Cylinders                      552     0.06
Model                          552     0.06
VehicleIntroDate               552     0.06
CapitalOutstanding               2     0.00


In [27]:
# Show object-type columns (often need special handling)
print("\n Object-Type Columns:\n")
print(df.select_dtypes(include='object').columns.tolist())



 Object-Type Columns:

['TransactionMonth', 'Citizenship', 'LegalType', 'Title', 'Language', 'Bank', 'AccountType', 'MaritalStatus', 'Country', 'Province', 'MainCrestaZone', 'SubCrestaZone', 'ItemType', 'VehicleType', 'make', 'Model', 'bodytype', 'VehicleIntroDate', 'AlarmImmobiliser', 'TrackingDevice', 'CapitalOutstanding', 'NewVehicle', 'WrittenOff', 'Rebuilt', 'Converted', 'CrossBorder', 'TermFrequency', 'ExcessSelected', 'CoverCategory', 'CoverType', 'CoverGroup', 'Section', 'Product', 'StatutoryClass', 'StatutoryRiskType']


In [28]:
# Calculate missing percentage per column
missing_percent = df.isna().mean() * 100

# Identify columns with > 50% missing data
cols_to_drop = missing_percent[missing_percent > 50].index.tolist()
print(f"Dropping columns with > 50% missing: {cols_to_drop}")

# Drop these columns
df.drop(columns=cols_to_drop, inplace=True)

# Confirm removal
print(f"Remaining columns: {df.columns.tolist()}")


Dropping columns with > 50% missing: ['CustomValueEstimate', 'WrittenOff', 'Rebuilt', 'Converted', 'CrossBorder']
Remaining columns: ['UnderwrittenCoverID', 'PolicyID', 'TransactionMonth', 'IsVATRegistered', 'Citizenship', 'LegalType', 'Title', 'Language', 'Bank', 'AccountType', 'MaritalStatus', 'Gender', 'Country', 'Province', 'PostalCode', 'MainCrestaZone', 'SubCrestaZone', 'ItemType', 'mmcode', 'VehicleType', 'RegistrationYear', 'make', 'Model', 'Cylinders', 'cubiccapacity', 'kilowatts', 'bodytype', 'NumberOfDoors', 'VehicleIntroDate', 'AlarmImmobiliser', 'TrackingDevice', 'CapitalOutstanding', 'NewVehicle', 'SumInsured', 'TermFrequency', 'CalculatedPremiumPerTerm', 'ExcessSelected', 'CoverCategory', 'CoverType', 'CoverGroup', 'Section', 'Product', 'StatutoryClass', 'StatutoryRiskType', 'TotalPremium', 'TotalClaims']


In [29]:
# List of categorical columns to fill with "Unknown" (adjust as needed)
cat_cols_to_fill = ['Bank', 'NewVehicle', 'AccountType', 'TrackingDevice', 'AlarmImmobiliser']

for col in cat_cols_to_fill:
    if col in df.columns:
        df[col] = df[col].fillna('Unknown')


In [34]:
for col in num_cols_to_impute:
    if col in df.columns:
        # Convert to numeric with coercion
        df[col] = pd.to_numeric(df[col], errors='coerce')
        
        # Check if all values are NaN
        if df[col].isna().all():
            print(f"Column '{col}' has all NaN values after conversion.")
        else:
            median_val = df[col].median()
            df[col] = df[col].fillna(median_val)


Column 'TermFrequency' has all NaN values after conversion.
Column 'ExcessSelected' has all NaN values after conversion.


In [35]:
cols_to_drop = ['TermFrequency', 'ExcessSelected']

# Drop columns if they exist in the DataFrame
df = df.drop(columns=[col for col in cols_to_drop if col in df.columns])

print("Dropped columns:", cols_to_drop)


Dropped columns: ['TermFrequency', 'ExcessSelected']


In [32]:
vehicle_features = ['Length', 'Width', 'Height', 'EngineSize', 'FuelTankCapacity', 
                   'GrossVehicleMass', 'Power', 'NumberOfDoors', 'NumberOfSeats', 
                   'NumberOfCylinder']

existing_vehicle_features = [col for col in vehicle_features if col in df.columns]

df.dropna(subset=existing_vehicle_features, inplace=True)

print(f"Rows remaining after dropping missing vehicle feature rows: {len(df)}")


Rows remaining after dropping missing vehicle feature rows: 1000098


# Final Check Code

In [36]:

# 1. Check total rows and columns
print(f"Dataset shape: {df.shape}")

# 2. Check for missing values (any columns with missing data left?)
missing_summary = df.isnull().sum()
missing_cols = missing_summary[missing_summary > 0]

if missing_cols.empty:
    print("No missing values remaining in the dataset.")
else:
    print("Columns with missing values:")
    print(missing_cols)

# 3. Check for columns with single unique value (could be dropped if no variance)
single_value_cols = [col for col in df.columns if df[col].nunique() <= 1]
if single_value_cols:
    print(f"Columns with single unique value (consider dropping): {single_value_cols}")
else:
    print("No columns with single unique value.")

# 4. Check data types summary
print("\nData types summary:")
print(df.dtypes.value_counts())

# 5. Summary statistics for numerical columns
print("\nNumerical columns summary statistics:")
print(df.describe())

# 6. Summary for categorical columns
cat_cols = df.select_dtypes(include=['object', 'category']).columns
print("\nCategorical columns unique value counts:")
for col in cat_cols:
    print(f"{col}: {df[col].nunique()} unique values")

# 7. Quick check for duplicates
duplicates_count = df.duplicated().sum()
print(f"\nNumber of duplicate rows: {duplicates_count}")

Dataset shape: (1000098, 44)
Columns with missing values:
mmcode              552
VehicleType         552
make                552
Model               552
Cylinders           552
cubiccapacity       552
kilowatts           552
bodytype            552
VehicleIntroDate    552
dtype: int64
Columns with single unique value (consider dropping): ['Language', 'Country', 'ItemType', 'StatutoryClass', 'StatutoryRiskType']

Data types summary:
object      28
float64     10
int64        4
bool         1
category     1
Name: count, dtype: int64

Numerical columns summary statistics:
       UnderwrittenCoverID      PolicyID    PostalCode        mmcode  \
count         1.000098e+06  1.000098e+06  1.000098e+06  9.995460e+05   
mean          1.048175e+05  7.956682e+03  3.020601e+03  5.487770e+07   
std           6.329371e+04  5.290039e+03  2.649854e+03  1.360381e+07   
min           1.000000e+00  1.400000e+01  1.000000e+00  4.041200e+06   
25%           5.514300e+04  4.500000e+03  8.270000e+02  6.00569

In [37]:
cols_to_drop = ['Language', 'Country', 'ItemType', 'StatutoryClass', 'StatutoryRiskType']
df = df.drop(columns=cols_to_drop)


In [38]:
df = df.drop_duplicates()
print(f"Shape after dropping duplicates: {df.shape}")


Shape after dropping duplicates: (1000024, 39)


In [39]:
vehicle_missing_cols = ['mmcode', 'VehicleType', 'make', 'Model', 'Cylinders', 'cubiccapacity', 'kilowatts', 'bodytype', 'VehicleIntroDate']
df = df.dropna(subset=vehicle_missing_cols)
print(f"Shape after dropping rows with missing vehicle info: {df.shape}")


Shape after dropping rows with missing vehicle info: (999472, 39)


In [40]:
invalid_rows = (df['TotalPremium'] < 0) | (df['TotalClaims'] < 0) | (df['CapitalOutstanding'] < 0)
print(f"Number of rows with negative TotalPremium, TotalClaims or CapitalOutstanding: {invalid_rows.sum()}")

df = df[~invalid_rows]
print(f"Shape after removing invalid negative values: {df.shape}")


Number of rows with negative TotalPremium, TotalClaims or CapitalOutstanding: 570
Shape after removing invalid negative values: (998902, 39)


In [41]:
print(f"Final dataset shape: {df.shape}")
print("Missing values remaining:")
print(df.isnull().sum()[df.isnull().sum() > 0])
print(f"Duplicate rows remaining: {df.duplicated().sum()}")


Final dataset shape: (998902, 39)
Missing values remaining:
Series([], dtype: int64)
Duplicate rows remaining: 0


In [ ]:

    print(col)


Remaining columns:
UnderwrittenCoverID
PolicyID
TransactionMonth
IsVATRegistered
Citizenship
LegalType
Title
Bank
AccountType
MaritalStatus
Gender
Province
PostalCode
MainCrestaZone
SubCrestaZone
mmcode
VehicleType
RegistrationYear
make
Model
Cylinders
cubiccapacity
kilowatts
bodytype
NumberOfDoors
VehicleIntroDate
AlarmImmobiliser
TrackingDevice
CapitalOutstanding
NewVehicle
SumInsured
CalculatedPremiumPerTerm
CoverCategory
CoverType
CoverGroup
Section
Product
TotalPremium
TotalClaims


In [47]:
# List all columns in the DataFrame
print(" All Columns in the DataFrame:\n")
print(df.columns.tolist())

 All Columns in the DataFrame:

['UnderwrittenCoverID', 'PolicyID', 'TransactionMonth', 'IsVATRegistered', 'Citizenship', 'LegalType', 'Title', 'Bank', 'AccountType', 'MaritalStatus', 'Gender', 'Province', 'PostalCode', 'MainCrestaZone', 'SubCrestaZone', 'mmcode', 'VehicleType', 'RegistrationYear', 'make', 'Model', 'Cylinders', 'cubiccapacity', 'kilowatts', 'bodytype', 'NumberOfDoors', 'VehicleIntroDate', 'AlarmImmobiliser', 'TrackingDevice', 'CapitalOutstanding', 'NewVehicle', 'SumInsured', 'CalculatedPremiumPerTerm', 'CoverCategory', 'CoverType', 'CoverGroup', 'Section', 'Product', 'TotalPremium', 'TotalClaims']


In [49]:
import os
print("Current working directory:", os.getcwd())
print("List of files/folders here:", os.listdir())
print("Does 'data' folder exist here?", os.path.isdir("data"))


Current working directory: c:\Users\HP\Desktop\car-insurance-project-analysis-1\notebooks
List of files/folders here: ['data_cleaning.ipynb', 'eda_analysis.ipynb']
Does 'data' folder exist here? False


In [50]:
# Save the cleaned DataFrame to a CSV file
cleaned_file_path = "cleaned_data.csv"
df.to_csv(cleaned_file_path, index=False)

print(f"Cleaned dataset saved to '{cleaned_file_path}'")


Cleaned dataset saved to 'cleaned_data.csv'


# Dataset Cleaning Summary

- **Initial shape:** (1,000,098, 44)
- **Dropped columns:** ['Language', 'Country', 'ItemType', 'StatutoryClass', 'StatutoryRiskType']
- **Dropped duplicates:** 74 rows
- **Dropped rows with missing vehicle info:** 552 rows
- **Dropped rows with invalid negative values in TotalPremium, TotalClaims, CapitalOutstanding**
- **Final shape:** (998,902, 39)
- **No missing values remaining**
- **No duplicate rows remaining**
